Cubical/Structures/QuotientRing.agda

```diff
@@ -0,0 +1,248 @@
+{-# OPTIONS --cubical --safe #-}
+
+module Cubical.Structures.QuotientRing where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.HLevels
+open import Cubical.HITs.SetQuotients.Base
+open import Cubical.HITs.SetQuotients.Properties
+open import Cubical.Structures.Ring
+open import Cubical.Structures.Ideal
+
+
+private
+  variable
+    ℓ : Level
+
+module _ (R′ : Ring {ℓ}) (I : ⟨ R′ ⟩  → hProp ℓ) (I-isIdeal : isIdeal R′ I) where
+  private
+    open ring-syntax R′
+    open isIdeal I-isIdeal
+    open ring-axioms R′
+    open calculations R′
+    R = ⟨ R′ ⟩
+
+  R/I : Type ℓ
+  R/I = R / (λ x y → x - y ∈ I)
+
+  private
+    -isDistributive : (x y : R) →  (- x) + (- y) ≡ - (x + y)
+    -isDistributive x y =
+      implicitInverse _ _
+           ((x + y) + ((- x) + (- y)) ≡⟨ sym (ring+-assoc _ _ _) ⟩
+            x + (y + ((- x) + (- y))) ≡⟨ cong
+                                           (λ u → x + (y + u))
+                                           (ring+-comm _ _) ⟩
+            x + (y + ((- y) + (- x))) ≡⟨ cong (λ u → x + u) (ring+-assoc _ _ _) ⟩
+            x + ((y + (- y)) + (- x)) ≡⟨ cong (λ u → x + (u + (- x)))
+                                              (ring+-rinv _) ⟩
+            x + (₀ + (- x))           ≡⟨ cong (λ u → x + u) (ring+-lid _) ⟩
+            x + (- x)                 ≡⟨ ring+-rinv _ ⟩
+            ₀ ∎)
+
+    translatedDifference :
+      ∀ (x a b : R)
+      → a - b ≡ (x + a) - (x + b)
+    translatedDifference x a b =
+                a - b                       ≡⟨ cong (λ u → a + u)
+                                                    (sym (ring+-lid _)) ⟩
+                (a + (₀ + (- b)))          ≡⟨ cong (λ u → a + (u + (- b)))
+                                                    (sym (ring+-rinv _)) ⟩
+                (a + ((x + (- x)) + (- b))) ≡⟨ cong (λ u → a + u)
+                                                    (sym (ring+-assoc _ _ _)) ⟩
+                (a + (x + ((- x) + (- b)))) ≡⟨ (ring+-assoc _ _ _) ⟩
+                ((a + x) + ((- x) + (- b))) ≡⟨ cong (λ u → u + ((- x) + (- b)))
+                                                    (ring+-comm _ _) ⟩
+                ((x + a) + ((- x) + (- b))) ≡⟨ cong (λ u → (x + a) + u)
+                                                    (-isDistributive _ _) ⟩
+                ((x + a) - (x + b)) ∎
+
+
+    homogenity : ∀ (x a b : R)
+                 → (a - b ∈ I)
+                 → (x + a) - (x + b) ∈ I
+    homogenity x a b p = subst (λ u → u ∈ I) (translatedDifference x a b) p
+
+
+    translate : R → R/I → R/I
```

Comment:
I'm not super fond of this name... What is even doing? I guess the intuition for the name is that it's some affine transformation like translation?

Reply:
I renamed it to "pre-+/I", because it is really just the addition on the quotient, before I continue to show, that it decends to the quotient. I also renamed some lemma about it to "pre-+/I-DescendsToQuotient". I guess it is clearer now, what things are.